### PR TITLE
Add message metadata

### PR DESCRIPTION
**Pull request by ben-challis**

### CODE REVIEW

File: src/Metadata/Metadata.php [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+     * Constructs metadata from an array, the array must contain the keys "Message-Name", "Correlation-Id" and
+     * "Causation-Id", containing their respective values.
+     *
+     * @param array<string, scalar> $metadata
+     *
+     * @throws \InvalidArgumentException If "Message-Name", "Causation-Id" and/or "Correlation-Id" are missing from
+     * the array, or are not string values.
+     * @throws InvalidMessageName If the "Message-Name" key is invalid.
+     * @throws InvalidUuidString If the "Message-Id" key is not a hex encoded UUID string.
+     * @throws InvalidUuidString If the "Correlation-Id" key is not a hex encoded UUID string.
```
Comment (ben-challis) [inline src/Metadata/Metadata.php:61]: This is not currently true for `CorrelationId`, need to fix it: https://github.com/Lendable/message/issues/39

File: src/Metadata/Metadata.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        );
+    }
+
+    public function without(string $key): self
+    {
+        $metadata = $this->flatten();
+
+        unset($metadata[$key]);
+
+        return self::fromArray($metadata);
```
Comment (ben-challis) [inline src/Metadata/Metadata.php:130]: Instead should probably do along the lines of:

```php
if (MetadataKey::tryFrom($key) instanceof Metadata) { 
    throw new \InvalidArgumentException('Cannot remove mandatory...'); 
}
```

File: src/Metadata/MetadataEnricher.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lendable\Message\Metadata;
+
+use Lendable\Message\Message;
+
+/**
+ * Metadata enrichers add new key/value pairs into an existing {@see Metadata}, providing a new instance of the
```
Comment (marmichalski) [inline src/Metadata/MetadataEnricher.php:10]:
```suggestion
 * Metadata enrichers add new key-value pairs into an existing {@see Metadata}, providing a new instance of the
```

File: src/Metadata/ChainEnricher.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lendable\Message\Metadata;
+
+use Lendable\Message\Message;
+
+/**
+ * Enriches metadata by utilising multiple other enrichers.
```
Comment (marmichalski) [inline src/Metadata/ChainEnricher.php:10]:
```suggestion
 * Enriches metadata by utilizing multiple other enrichers.
```

**Review by marmichalski — APPROVED** [review]

**Pull request merged**